llama_index/finetuning/embeddings/sentence_transformer.py

```diff
@@ -0,0 +1,171 @@
+"""Sentence Transformer Finetuning Engine."""
+
+from llama_index.embeddings.base import BaseEmbedding
+
+from typing import Dict, Any, List, Optional
+from pydantic import BaseModel
+
+from llama_index.schema import TextNode, MetadataMode
+from llama_index.llms.openai import OpenAI
+from llama_index.llms.base import LLM
+from llama_index.embeddings.utils import resolve_embed_model
+from llama_index.finetuning.types import BaseEmbeddingFinetuneEngine
+from tqdm import tqdm
+import uuid
+import re
+import json
+
+
+class EmbeddingQAFinetuneDataset(BaseModel):
+    """Embedding QA Finetuning Dataset."""
+
+    queries: Dict[str, str]  # dict id -> query
+    corpus: Dict[str, str]  # dict id -> string
+    relevant_docs: Dict[str, List[str]]  # query id -> list of doc ids
+
+    def save_json(self, path: str) -> None:
+        """Save json."""
+        with open(path, "w") as f:
+            json.dump(self.dict(), f, indent=4)
+
+    @classmethod
+    def from_json(cls, path: str) -> "EmbeddingQAFinetuneDataset":
+        """Load json."""
+        with open(path, "r") as f:
+            data = json.load(f)
+        return cls(**data)
+
+
+DEFAULT_QA_GENERATE_PROMPT_TMPL = """\
+Context information is below.
+
+---------------------
+{context_str}
+---------------------
+
+Given the context information and not prior knowledge.
+generate only questions based on the below query.
+
+You are a Teacher/ Professor. Your task is to setup \
+{num_questions_per_chunk} questions for an upcoming \
+quiz/examination. The questions should be diverse in nature \
+across the document. Restrict the questions to the \
+context information provided."
+"""
+
+
+# generate queries as a convenience function
+def generate_qa_embedding_pairs(
+    nodes: List[TextNode],
+    llm: Optional[LLM] = None,
+    qa_generate_prompt_tmpl: str = DEFAULT_QA_GENERATE_PROMPT_TMPL,
+    num_questions_per_chunk: int = 2,
+) -> EmbeddingQAFinetuneDataset:
+    """Generate examples given a set of nodes."""
+    node_dict = {
+        node.node_id: node.get_content(metadata_mode=MetadataMode.NONE)
+        for node in nodes
+    }
+
+    llm = OpenAI(model="gpt-3.5-turbo")
+
+    queries = {}
+    relevant_docs = {}
+    for node_id, text in tqdm(node_dict.items()):
```

Comment:
as a followup, I'd like to properly implement batch LLM calls, so this doesn't take so long